components/eam/src/physics/cam/modal_aer_opt.F90

```diff
@@ -1421,64 +1201,87 @@ subroutine calc_refin_complex (ncol, ilw,                   & ! in
+       if (lwsw=='lw') then
+          crefin(icol) = crefin(icol) + watervol(icol)*crefwlw(ilwsw)
+          if (wetvol(icol) > small_value_40) crefin(icol) = crefin(icol)/wetvol(icol)
+       elseif (lwsw=='sw') then
+          crefin(icol) = crefin(icol) + watervol(icol)*crefwsw(ilwsw)
+          crefin(icol) = crefin(icol)/max(wetvol(icol), small_value_60)
+       endif
+
+       refr(icol) = real(crefin(icol))
+       refi(icol) = aimag(crefin(icol))
+    enddo
 
 end subroutine calc_refin_complex
 
 !===================================================================
-subroutine check_error_warning(icol, kk, mm, ilw, nspec,list_idx, & ! in
+subroutine check_error_warning(lwsw, icol, kk, mm, ilwsw, nspec,list_idx,   & ! in
                         dopaer, pabs, dryvol, wetvol, watervol, crefin,cabs,& ! in
-                        specdens, specrefindex, volf, & ! in
-                        nerr_dopaer) ! inout
+                        specdens, specrefindex, specvol, & ! in
+                        nerr_dopaer,   & ! inout
+                        pext, specpext ) ! optional in
     !------------------------------------------------------------
     ! check and writeout error and warning message
     !------------------------------------------------------------
 
    implicit none
-   integer,intent(in) :: icol, kk, mm, ilw, nspec, list_idx  ! indices
+   character(len=2), intent(in) :: lwsw   ! indicator if this is lw or sw
+   integer,intent(in) :: icol, kk, mm, ilwsw, nspec, list_idx  ! indices
    real(r8),intent(in) :: dopaer    ! aerosol optical depth in layer
    real(r8),intent(in) :: pabs      ! parameterized specific absorption [m2/kg]
    real(r8),intent(in) :: dryvol(:)    ! volume concentration of aerosol mode [m3/kg]
    real(r8),intent(in) :: wetvol(:)    ! volume concentration of wet mode [m3/kg]
    real(r8),intent(in) :: watervol(:)  ! volume concentration of water in each mode [m3/kg]
    complex(r8),intent(in) :: crefin(:) ! complex refractive index
    real(r8),intent(in) :: cabs(:,:)
-   real(r8),intent(in) :: volf(:,:)   ! volume fraction of insoluble aerosol [fraction]
-   real(r8),intent(in) :: specdens(:) ! species density [kg/m3]
+   real(r8),intent(in) :: specvol(:,:) ! volume concentration of aerosol specie [m3/kg]
+   real(r8),intent(in) :: specdens(:)  ! species density [kg/m3]
    complex(r8), intent(in) :: specrefindex(:, :)     ! species refractive index
    integer, intent(inout) :: nerr_dopaer    ! total number of error times
+   real(r8),intent(in),optional :: pext         ! only write out for sw
+   real(r8),intent(in),optional :: specpext     ! only write out for sw
 
    integer :: ll
    integer, parameter :: nerrmax_dopaer=1000
+   integer, parameter :: small_value_neg = -1.e-10_r8
+
+    if ((lwsw /= 'lw') .and. (lwsw /= 'sw')) then
+        write(err_msg,*)'check_error_warning is called with ', lwsw, ', it should be called with either lw or sw'
+        call endrun()
+    endif
```

Comment:
Are you able to compile this code? I don't see `err_msg` defined anywhere. Also, the `endrun` call should have `err_msg` as an argument:
```
call endrun(err_msg)
```

Reply:
Yeah, `endrun` takes an "optional" string argument. It should work even without an argument. In my suggestion, I was writing the error message to a variable `err_msg` as a character string. This needs to be sent to `endrun` call otherwise the code will not print this error message if it fails.